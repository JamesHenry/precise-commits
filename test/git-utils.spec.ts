import { join } from 'path';

import {
  TMP_DIRECTORY_PATH,
  destroyTmpDirectory,
  readFixtures,
  getTmpFilePath,
  createTmpDirectoryAndInitialiseGit,
  prepareFixtureInTmpDirectory,
} from './test-utils';
import {
  getDiffForFile,
  resolveNearestGitDirectory,
  getStagedModifiedFiles,
} from '../src/git-utils';

const fixtures = readFixtures();

describe('git-utils', () => {
  describe('resolveNearestGitDirectory()', () => {
    beforeEach(createTmpDirectoryAndInitialiseGit);

    it('should find the nearest .git directory to the given file by traversing up the parent hierarchy', () => {
      /**
       * This current spec file should use the repo's root .git
       */
      expect(resolveNearestGitDirectory(__dirname)).toEqual(
        join(__dirname, '../.git'),
      );
      /**
       * The tmp file should resolve to its own .git directory within the tmp directory
       */
      expect(resolveNearestGitDirectory(TMP_DIRECTORY_PATH)).toEqual(
        join(TMP_DIRECTORY_PATH, '.git'),
      );
    });

    afterEach(destroyTmpDirectory);
  });

  describe('getDiffForFile()', () => {
    beforeEach(createTmpDirectoryAndInitialiseGit);

    fixtures.forEach(fixture => {
      it(fixture.fixtureName, () => {
        prepareFixtureInTmpDirectory(fixture);
        const diff = getDiffForFile(getTmpFilePath());
        expect(diff).toMatchSnapshot();
      });
    });

    afterEach(destroyTmpDirectory);
  });

  describe('getStagedModifiedFiles()', () => {
    beforeEach(createTmpDirectoryAndInitialiseGit);

    fixtures.forEach(fixture => {
      it(fixture.fixtureName, () => {
        prepareFixtureInTmpDirectory(fixture);
        const fileNames = getStagedModifiedFiles(TMP_DIRECTORY_PATH);
        expect(fileNames).toEqual([`tmp${fixture.fileExtension}`]);
      });
    });

    afterEach(destroyTmpDirectory);
  });
});
