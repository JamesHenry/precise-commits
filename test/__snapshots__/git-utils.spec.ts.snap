// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`git-utils getDiffForFile() brand-new-jsx-file 1`] = `
"diff --git a/tmp.jsx b/tmp.jsx
new file mode 100644
index 0000000..ae72777
--- /dev/null
+++ b/tmp.jsx
@@ -0,0 +1,21 @@
+function HelloWorld({greeting = \\"hello\\", greeted = '\\"World\\"', silent = false, onMouseOver,}) {
+
+    if(!greeting){return null};
+  
+       // TODO: Don't use random in render
+    let num = Math.floor (Math.random() * 1E+7).toString().replace(/\\\\.\\\\d+/ig, \\"\\")
+  
+    return <div className='HelloWorld' title={\`You are visitor number \${ num }\`} onMouseOver={onMouseOver}>
+  
+      <strong>{ greeting.slice( 0, 1 ).toUpperCase() + greeting.slice(1).toLowerCase() }</strong>
+      {greeting.endsWith(\\",\\") ? \\" \\" : <span style={{color: '\\\\grey'}}>\\", \\"</span> }
+      <em>
+      { greeted }
+      </em>
+      { (silent)
+        ? \\".\\"
+        : \\"!\\"}
+  
+      </div>;
+  
+  }
\\\\ No newline at end of file"
`;

exports[`git-utils getDiffForFile() empty-javascript-file 1`] = `
"diff --git a/tmp.js b/tmp.js
new file mode 100644
index 0000000..3772dd0
--- /dev/null
+++ b/tmp.js
@@ -0,0 +1,2 @@
+var greeting = 'hello'
+        var audience = 'world'
\\\\ No newline at end of file"
`;

exports[`git-utils getDiffForFile() multiple-updated-lines-javascript-file 1`] = `
"diff --git a/tmp.js b/tmp.js
index 50d7ab7..2b98ecd 100644
--- a/tmp.js
+++ b/tmp.js
@@ -2 +2 @@ var     untouchedUglyLines1 = ''
-function TouchedFunctionSignature({greeting = \\"Hello\\", greeted = '\\"World\\"', silent = false,}) {
+function TouchedFunctionSignature({greeting = \\"Hello\\", greeted = '\\"Mars\\"', silent = false,}) {
@@ -5,0 +6 @@ function TouchedFunctionSignature({greeting = \\"Hello\\", greeted = '\\"World\\"', sile
+      const newlyAddedUglyLine = true
\\\\ No newline at end of file"
`;

exports[`git-utils getDiffForFile() single-updated-line-typescript 1`] = `
"diff --git a/tmp.ts b/tmp.ts
index 30011dc..1f5fc71 100644
--- a/tmp.ts
+++ b/tmp.ts
@@ -2 +2 @@ const a = true
-const b = false
\\\\ No newline at end of file
+const b = true
\\\\ No newline at end of file"
`;
